alibi_detect/saving/saving.py

```diff
@@ -49,23 +50,63 @@ def save_detector(detector: Detector, filepath: Union[str, os.PathLike], legacy:
     if detector_name not in [detector.__name__ for detector in get_args(Detector)]:
         raise NotImplementedError(f'{detector_name} is not supported by `save_detector`.')
 
-    # check if path exists
+    # Get a list of all existing files in `filepath` (so we know what not to cleanup if an error occurs)
     filepath = Path(filepath)
-    if not filepath.is_dir():
-        logger.warning('Directory {} does not exist and is now created.'.format(filepath))
-        filepath.mkdir(parents=True, exist_ok=True)
-
-    # If a drift detector, wrap drift detector save method
-    if hasattr(detector, 'get_config') and not legacy:
-        _save_detector_config(detector, filepath)
+    orig_files = set(filepath.iterdir())
+
+    # Saving is wrapped in a try, with cleanup in except. To prevent a half-saved detector remaining upon error.
+    # Cleanup is only performed if `filepath` was a new directory, so that we avoid unintentionally deleting the
+    # `filepath` directory if it contains other items unrelated to saving.
+    # TODO - we could delete in a more targeted fashion if we maintain a manifest of all possible files/dirs written
+    try:
+        # Create directory if it doesn't exist
+        if not filepath.is_dir():
+            logger.warning('Directory {} does not exist and is now created.'.format(filepath))
+            filepath.mkdir(parents=True, exist_ok=True)
+
+        # If a drift detector, wrap drift detector save method
+        if hasattr(detector, 'get_config') and not legacy:
+            _save_detector_config(detector, filepath)
+
+        # Otherwise, save via the previous meta and state_dict approach
+        else:
+            save_detector_legacy(detector, filepath)
 
-    # Otherwise, save via the previous meta and state_dict approach
-    else:
-        save_detector_legacy(detector, filepath)
+    except Exception as error:
+        _cleanup_filepath(orig_files, filepath)
+        print('Saving failed due to the following error. Files/directories have been cleaned up.')
+        raise error
```

Comment:
Why not put the message in an error, i.e. something like `raise SomeError('text') from error`.

Reply:
That'd be much nicer, don't know why I did the above. Will change!

Reply:
Replaced with 

```python
raise RuntimeError(f'Saving failed. The save directory {filepath} has been cleaned.') from error
```

Also removed old inaccurate comments regarding the cleanup logic.